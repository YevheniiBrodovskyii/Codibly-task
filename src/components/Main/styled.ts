import styled from 'styled-components'

export const ErrorMessage = styled.div`
	min-height: 670px;
	padding-top: 100px;
	text-align: center;
	font-size: 30px;
`
