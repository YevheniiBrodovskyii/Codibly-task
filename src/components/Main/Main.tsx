import { FC, useEffect, useState } from 'react'
import { useSearchParams } from 'react-router-dom'

import { useAppDispatch } from '../../hook'
import { getProducts } from '../../store/ProductsSlice'

import Filter from '../Filter/Filter'
import Pagination from '../Pagination/Pagination'
import Products from '../Products/Products'

import { ErrorMessage } from './styled'

const Main: FC = () => {
	const dispatch = useAppDispatch()

	const [errorId, setErrorId] = useState<string>('')
	const [errorMessage, setErrorMessage] = useState<string>('')

	const [searchParams, setSearchParams] = useSearchParams()

	const page = searchParams.get('page') || '1'
	const filterId = searchParams.get('id') || ''

	const setPage = (page: string) =>
		setSearchParams(
			(params) => {
				params.set('page', page)
				return params
			},
			{ replace: true }
		)

	const setFilterId = (id: string) =>
		setSearchParams(
			(params) => {
				params.set('id', id)
				return params
			},
			{ replace: true }
		)

	useEffect(() => {
		async function fetchProducts(id?: string, page = '1') {
			const itemsPerPage = 5
			const url = id
				? `https://reqres.in/api/products/${id}`
				: `https://reqres.in/api/products?per_page=${itemsPerPage}&page=${page}`

			const requestOptions = {
				method: 'GET',
				headers: { 'Content-Type': 'application/json' },
			}

			try {
				const response = await fetch(url, requestOptions)
				const data = await response.json()

				if (response.ok) {
					setErrorId('')
					setErrorMessage('')

					if (id) {
						dispatch(
							getProducts({
								page: 1,
								per_page: 1,
								total: 1,
								total_pages: 1,
								data: [data.data],
							})
						)
						setSearchParams({ page, id }, { replace: true })
					} else {
						dispatch(getProducts(data))
						setSearchParams({ page }, { replace: true })
					}
				} else {
					const error = data?.message || response.status
					setErrorId(error.toString())
					return Promise.reject(error)
				}
			} catch (error: any) {
				setErrorId(error.toString())
				console.error('There was an error!', error)
			}
		}

		fetchProducts(filterId, page)
	}, [filterId, page])
	return (
		<div>
			<Filter setFilterId={setFilterId} filterId={filterId} />
			{errorId ? (
				<ErrorMessage>
					{errorId === '404'
						? 'Product not found'
						: `${errorId}: ${errorMessage}`}
				</ErrorMessage>
			) : (
				<Products />
			)}
			<Pagination setPage={setPage} />
		</div>
	)
}

export default Main
