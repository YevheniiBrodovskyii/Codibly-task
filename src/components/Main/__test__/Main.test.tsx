import Main from '../Main'

import { RenderResult } from '@testing-library/react'

import { renderWithProviders } from '../../../utils/test-utils'
import { BrowserRouter } from 'react-router-dom'

describe('Main tests', () => {
	let component: RenderResult
	beforeEach(() => {
		component = renderWithProviders(
			<BrowserRouter>
				<Main />
			</BrowserRouter>
		)
	})

	it('should match snapshot on initial render', () => {
		expect(component).toMatchSnapshot()
	})

	// it('should trigger fetch arrow button', async () => {

	// })
})
