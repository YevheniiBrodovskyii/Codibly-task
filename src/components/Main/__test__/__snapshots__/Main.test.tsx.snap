// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main tests should match snapshot on initial render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="sc-bcXHqe jnCFJK"
        >
          <input
            class="sc-gswNZR hAgmCS"
            data-testid="filter"
            placeholder="Search by id"
            type="number"
            value=""
          />
        </div>
        <div
          class="sc-fEXmlR iFeXHN"
        >
          <div
            class="sc-eDWCr jhZGKn"
            color=""
          >
            <h3
              class="sc-bqWxrE grVcgM"
            >
              Item id: 
              0
            </h3>
            <h1
              class="sc-ksBlkl gYWRFf"
            />
            <span
              class="sc-fnGiBr kaJPKD"
            >
              Click for more...
            </span>
            <h2
              class="sc-hBxehG dvtxCh"
            >
              0
            </h2>
          </div>
        </div>
        <div
          class="sc-dkrFOg bSuhkq"
        >
          <button
            class="sc-hLBbgP jcMiJo"
            disabled=""
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
              data-testid="ArrowBackIosIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
              />
            </svg>
          </button>
          <button
            class="sc-hLBbgP jcMiJo"
            data-testid="pagination-right-arrow"
            disabled=""
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
              data-testid="ArrowForwardIosIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="sc-bcXHqe jnCFJK"
      >
        <input
          class="sc-gswNZR hAgmCS"
          data-testid="filter"
          placeholder="Search by id"
          type="number"
          value=""
        />
      </div>
      <div
        class="sc-fEXmlR iFeXHN"
      >
        <div
          class="sc-eDWCr jhZGKn"
          color=""
        >
          <h3
            class="sc-bqWxrE grVcgM"
          >
            Item id: 
            0
          </h3>
          <h1
            class="sc-ksBlkl gYWRFf"
          />
          <span
            class="sc-fnGiBr kaJPKD"
          >
            Click for more...
          </span>
          <h2
            class="sc-hBxehG dvtxCh"
          >
            0
          </h2>
        </div>
      </div>
      <div
        class="sc-dkrFOg bSuhkq"
      >
        <button
          class="sc-hLBbgP jcMiJo"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="ArrowBackIosIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
            />
          </svg>
        </button>
        <button
          class="sc-hLBbgP jcMiJo"
          data-testid="pagination-right-arrow"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="ArrowForwardIosIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
